utils/src/main/java/software/amazon/awssdk/utils/async/AbstractFlatteningSubscriber.java

```diff
@@ -0,0 +1,274 @@
+        handleStateUpdate();
+    }
+
+    /**
+     * Increment the downstream demand by the provided value, accounting for overflow.
+     */
+    private void addDownstreamDemand(long l) {
+        if (l > 0) {
+            downstreamDemand.getAndUpdate(current -> {
+                long newValue = current + l;
+                return newValue >= 0 ? newValue : Long.MAX_VALUE;
+            });
+        } else {
+            log.error(() -> "Demand " + l + " must not be negative.");
+            upstreamSubscription.cancel();
+            onError(new IllegalArgumentException("Demand must not be negative"));
+        }
+    }
+
+    /**
+     * This is invoked after each downstream request or upstream onNext, onError or onComplete.
+     */
+    private void handleStateUpdate() {
+        do {
+            // Anything that happens after this if statement and before we set handlingStateUpdate to false is guaranteed to only
+            // happen on one thread. For that reason, we should only invoke onNext, onComplete or onError within that block.
+            if (!handlingStateUpdate.compareAndSet(false, true)) {
+                return;
+            }
+
+            try {
+                // If we've already called onComplete or onError, don't do anything.
+                if (terminalCallMadeDownstream) {
+                    return;
+                }
+
+                // Call onNext, onComplete and onError as needed based on the current subscriber state.
+                handleOnNextState();
+                handleUpstreamDemandState();
+                handleOnCompleteState();
+                handleOnErrorState();
+            } catch (Error e) {
+                throw e;
+            } catch (Throwable e) {
+                log.error(() -> "Unexpected exception encountered that violates the reactive streams specification. Attempting "
+                                + "to terminate gracefully.", e);
+                upstreamSubscription.cancel();
+                onError(e);
+            } finally {
+                handlingStateUpdate.set(false);
+            }
+
+            // It's possible we had an important state change between when we decided to release the state update flag, and we
+            // actually released it. If that seems to have happened, try to handle that state change on this thread, because
+            // another thread is not guaranteed to come around and do so.
+        } while (onNextNeeded() || upstreamDemandNeeded() || onCompleteNeeded() || onErrorNeeded());
+    }
+
+    /**
+     * Fulfill downstream demand by flushing
+     */
+    private void handleOnNextState() {
+        while (onNextNeeded() && !onErrorNeeded()) {
+            fulfillDownstreamDemand();
+        }
+    }
+
+    /**
+     * Returns true if we need to call onNext downstream. If this is executed outside the handling-state-update condition, the
+     * result is subject to change.
+     */
+    protected boolean onNextNeeded() {
```

Comment:
See my previous comment, suggesting template methods to allow subclasses to add additional checks. Can we make those methods that are intended to be overridden by subclasses package private?